c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -282,13 +282,102 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                 */
                 Label::default()
             }),
-            Rvalue::Aggregate(ref kind, ref _ops) => match **kind {
+            Rvalue::Aggregate(ref kind, ref ops) => match **kind {
                 AggregateKind::Array(..) => {
                     let ty = rv.ty(self.local_decls, *self.ltcx);
                     // TODO: create fresh origins for all pointers in `ty`, and generate subset
                     // relations between the regions of the array and the regions of its elements
                     self.ltcx.label(ty, &mut |_ty| Label::default())
                 }
+                AggregateKind::Adt(adt_did, ..) => {
+                    /*
+                        Generic types are not yet supported because of situations such as the
+                        following:
+
+                        ```rust
+                        struct S<T> {
+                            s: T
+                        };
+
+                        fn foo() {
+                            let s = S<&'1 u32> { s: &'1 0 }
+                        }
+                        ```
+
+                        Here the `FieldMetadata` for `S:s` has an empty `origin_args` slice
+                        because these are gathered before the lifetime parameters are resolved
+                        by traversing the struct definition. Additionally, the label for the
+                        `Operand` corresponding to `s` has an `Origin('1)` (because it is
+                        a reference). Because the `OriginArgs` are missing for the field, the
+                        labeled types for the field operand and for the field declaration do not
+                        have the same shape, and so `do_assign` cannot create subset relations.
+
+                        Additionally, the regions referenced within the operand types are
+                        erased, and so it also would not be possible to know which ADT
+                        OriginParams those correspond to because there would need to be a
+                        non-erased region to compare with.
```

Comment:
@spernsteiner how would you feel about me creating an issue for supporting this and generics? AFAICT this isn't going to be a consideration until we want to incrementally lift (I don't see any transpiled code that uses generics in lighttpd)

Reply:
Putting off generics for now is fine, but I think having a separate function for computing the field type would still make the logic below much clearer.

Reply:
I'm a bit confused about the refactoring feedback. I can imagine a way to break out the substitution into its own function but don't see the vision of refactoring specifically to return a relabeled field type; this block doesn't relabel field types and performs substitution on the RHS of the assignment, not the LHS / `foo` which should already have the appropriate `Origins` from the labeling that occurs in `visit_place` in `adt_func`. What am I missing?

Reply:
I didn't look at the implementation here in detail, but I was expecting it to follow the same general structure as in `dataflow`, where we effectively unroll the aggregate assignment into a series of field assignments.  For example, we would treat `s = S { x: a, y: b }` as if it were `s.x = a; s.y = b;`.  So the logic for each field would be:
1. From the struct type (which in this case I think is `expect_ty`), obtain the type of the field
2. Get the type of the value being stored in the field (here it's `op_label` - this should maybe be called `op_lty`, since it's the whole `LTy`, not just the label)
3. Pseudo-assign the value type into the field type

It sounds like what you're describing is the other way around:

>performs substitution on the RHS of the assignment, not the LHS / foo

It sounds like the idea here is to handle `s = S { x: a, y: b }` by building up an `LTy` for `S { x: a, y: b }` based on the `LTy`s for `a` and `b` and then pseudo-assigning that `LTy` into the type of `s`.  But I don't think this is possible in general.  Suppose we have the following setup:

```Rust
struct S<'a> {
    x: &'a i32,
    y: &'a u8,
}
let a: &'o1 i32 = ...;
let b: &'o2 u8 = ...;
let s: S<'o3> = S { x: a, y: b };
```
There's no single origin to pick for `'a` in the type of `S { x: a, y: b }` - picking `S<'o1>` means ignoring the interaction with `'o2` and vice versa.

Reply:
> There's no single origin to pick for 'a in the type of S { x: a, y: b } - picking S<'o1> means ignoring the interaction with 'o2 and vice versa.

You're right, this is a bug. But `visit_statement` calls `do_assign` on the LHS and RHS, which also assigns the `Origins` of the lifetime parameters. It seems like the concrete origins for the `Aggregate` literal on the RHS need to be present, so how can I reconcile that with the sort of example you posted above where the fields have two different concrete origins but there is only one lifetime parameter for the struct? Should there be a different origin `'o3` that outlives `'o1` and `'o2`? Should that sort of assignment not happen at all?

I realized there was never really any assignment between fields of the LHS and RHS, so I've factored out the function and called the assignment in 249d6a55.

Reply:
One option is to take `expect_lty` to be the type of the struct literal.  Then you would use that as the base for the field projections, pseudo-assign `a` into `proj(expect_lty, x)` and `b` into `proj(expect_lty, y)`, and finally return `expect_lty` as the overall type of the rvalue.

>Should there be a different origin `'oX` that outlives `'o1` and `'o2`?

This is the other option, which matches what we do in `dataflow` with `rvalue_tys`.  We would label each `Rvalue` up front, including generating fresh `Origin`s for all its real and hypothetical region arguments.  Then we would pseudo-assign `a` into `proj(rvalue_lty, x)` etc, and return `rvalue_lty` to be pseudo-assigned into the LHS type (which is equal to `expect_lty`).

However, I'm not sure if this extra indirection is necessary here.  We use it in `dataflow` because it makes it easier to attach casts to the individual field values later on, but for `borrowck` we don't have such a requirement.  So it's probably okay to go with the first option unless you actually run into trouble with that approach.

Reply:
Ok, I've pushed option 1 in 9b9dc7a

Reply:
And made a small change in 1e8d88d